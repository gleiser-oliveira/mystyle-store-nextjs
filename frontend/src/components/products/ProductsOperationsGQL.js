import gql from 'graphql-tag';
import { perPage } from '../../config';

export const ALL_PRODUCTS_QUERY = gql`
    query ALL_PRODUCTS_QUERY($skip: Int = 0, $first: Int = ${perPage}) {
        products(first: $first, skip: $skip, orderBy: createdAt_DESC) {
            id
            title
            price
            description
            image
            largeImage
            user {
                id
            }
        }
    }
`;

export const SINGLE_PRODUCT_QUERY = gql`
    query SINGLE_PRODUCT_QUERY($id: ID!) {
        product(where: { id: $id }) {
            id
            title
            description
            price
            image
            largeImage
            user {
                id
            }
        }
    }
`;

export const CREATE_PRODUCT_MUTATION = gql`
    mutation CREATE_PRODUCT_MUTATION (
        $title: String!
        $description: String!
        $price: Int!
        $image: String
        $largeImage: String
    ) {
        createProduct (
            title: $title 
            description: $description 
            price: $price
            image: $image
            largeImage: $largeImage
        ) {
            id
        }
    }
`;

export const UPDATE_PRODUCT_MUTATION = gql`
    mutation UPDATE_PRODUCT_MUTATION (
        $id: ID!
        $title: String
        $description: String
        $price: Int
    ) {
        updateProduct (
            id: $id
            title: $title 
            description: $description 
            price: $price
        ) {
            id
        }
    }
`;

export const DELETE_PRODUCT_MUTATION = gql`
    mutation DELETE_PRODUCT_MUTATION($id: ID!) {
        deleteProduct(id: $id) {
            id
        }
    }
`;

export const PRODUCTS_PAGINATION_QUERY = gql`
  query PRODUCTS_PAGINATION_QUERY {
    productsConnection {
      aggregate {
        count
      }
    }
  }
`;
