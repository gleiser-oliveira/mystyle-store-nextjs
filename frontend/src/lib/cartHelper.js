export const numberOfProducts = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.quantity), 0);
}

export const calcTotalCartPrice = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.product.price * b.quantity), 0);
}

export const formatPrice = (priceInCents) => {
    const options = {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
    };
    // if its a whole, dollar amount, leave off the .00
    if (priceInCents % 100 === 0)
        options.minimumFractionDigits = 0;

    const formatter = new Intl.NumberFormat('en-US', options);
    return formatter.format(priceInCents / 100);
}