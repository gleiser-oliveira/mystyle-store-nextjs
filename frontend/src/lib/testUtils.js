import casual from 'casual';
import { act } from 'react-dom/test-utils';
import { CURRENT_USER_QUERY } from '../components/authentication/AuthenticationOperationsGQL';

// seeded for constant results
casual.seed(123);

const fakeUser = overrides => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    email: casual.email,
    permissions: ['USER', 'PRODUCT_CREATE'],
    cart: [],
    ...overrides,
});

const fakeUser2 = () => ({
    __typename: 'User',
    id: '2222',
    name: casual.name,
    email: casual.email,
    permissions: ['USER'],
    orders: [],
    cart: [],
});

const fakeAdminUser = () => ({
    __typename: 'User',
    id: '1111',
    name: casual.name,
    email: casual.email,
    permissions: ['ADMIN', 'PRODUCT_CREATE'],
    orders: [],
    cart: [],
});

const fakeProduct = () => ({
    __typename: 'Product',
    id: 'QWERTY0123',
    title: 'A Fake Product',
    price: 5000,
    description: 'This is a fake product for testing.',
    image: 'product.jpg',
    largeImage: 'largeProduct.jpg',
    user: fakeUser(),
});

const fakeProduct2 = () => ({
    __typename: 'Product',
    id: 'AS012',
    title: 'Another Fake Product',
    price: 2999,
    description: 'This is another fake product for testing.',
    image: 'product2.jpg',
    largeImage: 'bigProduct2.jpg',
    user: fakeUser(),
});

const fakeOrderProduct = () => ({
    __typename: 'OrderProduct',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
});

const fakeOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    products: [fakeOrderProduct(), fakeOrderProduct()],
    createdAt: 'May 1, 2020 00:00:00',
    user: fakeUser(),
});


const fakeCartProduct = overrides => ({
    __typename: 'CartProduct',
    id: 'prod123',
    quantity: 2,
    product: fakeProduct(),
    user: fakeUser(),
    ...overrides,
});

const fakeCartProduct2 = overrides => ({
    __typename: 'CartProduct',
    id: 'prod312',
    quantity: 2,
    product: fakeProduct2(),
    user: fakeUser(),
    ...overrides,
});

const cartUser = fakeUser();
const fakeCart = () => ([
    fakeCartProduct({user: cartUser}),
    fakeCartProduct2({user: cartUser}),
]);

// Fake LocalStorage
class LocalStorageMock {
    constructor() {
        this.store = {};
    }

    clear() {
        this.store = {};
    }

    getItem(key) {
        return this.store[key] || null;
    }

    setItem(key, value) {
        this.store[key] = value.toString();
    }

    removeItem(key) {
        delete this.store[key];
    }
}

// mocks
const signedInMocks = (user = fakeUser) => { return [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: user(),
            }
        }
    }
]};

const notSignedInMocks = () => ([
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: null,
            }
        }
    }
]);

const updateWrapper = async (wrapper, ms = 0) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, ms));
        wrapper.update();
    });
}

export {
    LocalStorageMock,
    fakeProduct,
    fakeProduct2,
    fakeUser,
    fakeUser2,
    fakeAdminUser,
    fakeCartProduct,
    fakeCartProduct2,
    fakeOrder,
    fakeCart,
    fakeOrderProduct,
    signedInMocks,
    notSignedInMocks,
    updateWrapper,
};
