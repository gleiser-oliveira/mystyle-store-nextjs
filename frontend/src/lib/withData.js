import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from './ClientGQL/LocalGQL';
import { DOCKER_BACKEND_URL, BROWSER_TO_BACKEND_URL } from '../config';

function createClient({ headers }) {
    // if using docker, the URI to the Yoga server must be either:
    // DOCKER_BACKEND_URL = the name of the backend server set in the docker network (used in SSR fetches)
    // BROWSER_TO_BACKEND_URL = the URL of the backend server that can be accessed from the client (used in the client's browser fetches)
    const isInsideFrontendServer = typeof window === 'undefined';
    const uri = isInsideFrontendServer ? DOCKER_BACKEND_URL : BROWSER_TO_BACKEND_URL;
    return new ApolloClient({
        uri: uri,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include',
                },
                headers,
            });
        },
        // local data
        clientState: {
            resolvers: {
                Mutation: {
                    toggleCart(_, variables, client) {
                        const { cartOpen } = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        // toggle the state
                        const data = {
                            data: { cartOpen: !cartOpen },
                        };
                        client.cache.writeData(data);
                        return data;
                    },
                },
                Query: {
                    localState(_, variables, client) {
                        const localState = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        return localState;
                    }
                }
            },
            defaults: {
                cartOpen: false,
            }
        }
    });
}

// TODO: check getDataFromTree warning - apollo
export default withApollo(createClient);