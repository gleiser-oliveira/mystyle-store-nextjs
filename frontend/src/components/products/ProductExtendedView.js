import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Query } from 'react-apollo';
import Loader from 'react-loader-spinner';
import Head from 'next/head';
import AddToCartButton from '../cart/AddToCartButton';
import Image from '../image/Image';
import ErrorMessage from '../error/ErrorMessage';
import { Center } from '../theme/generalStyles';
import { formatPrice } from '../../lib/cartHelper';
import { SINGLE_PRODUCT_QUERY } from './ProductsOperationsGQL';
import { ProductExtendedViewStyle, PriceTag } from './styles';

class ProductExtendedView extends Component {
    render() {
        return (
            <Query query={ SINGLE_PRODUCT_QUERY } variables={{ id: this.props.id }}>
                {(payload) => {
                    if (payload.error)
                        return (
                            <Center>
                                <ErrorMessage error={payload.error} />
                            </Center>
                        );
                    if (payload.loading)
                        return (
                            <ProductExtendedViewStyle>
                                <div className='cell'>
                                    <Center>
                                        <Loader type="TailSpin" visible={true} />
                                    </Center>
                                </div>
                            </ProductExtendedViewStyle>
                        );
                    if (!payload.data || !payload.data.product)
                        return (
                            <Center>
                                <p>No Product Found for {this.props.id}</p>
                            </Center>
                        );
                    const product = payload.data.product;
                    return (
                        <ProductExtendedViewStyle>
                            <Head>
                                <title>My Style | {product.title}</title>
                            </Head>
                            <div>
                                <div className='cell'>
                                    <div className='relative'>
                                        <PriceTag>{formatPrice(product.price)}</PriceTag>
                                        <Image src={product.largeImage} alt={product.title} />
                                        {/* <img src={product.largeImage} alt={product.title} /> */}
                                    </div>
                                </div>
                            </div>
                            <div className="details">
                                <h2>Viewing: {product.title}</h2>
                                <div>
                                    <h3>About this product:</h3>
                                    <p>{product.description}</p>
                                </div>
                                <AddToCartButton productId={this.props.id}/>
                            </div>
                        </ProductExtendedViewStyle>
                    );
                }}
            </Query>
        );
    }
}

ProductExtendedView.propTypes = {
    id: PropTypes.string,
};

export default ProductExtendedView;
