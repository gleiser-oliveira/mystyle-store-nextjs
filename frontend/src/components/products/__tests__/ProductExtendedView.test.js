import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';
import ProductExtendedView from '../ProductExtendedView';
import { SINGLE_PRODUCT_QUERY } from '../ProductsOperationsGQL';
import { fakeProduct } from '../../../lib/testUtils';

describe('Testing <ProductExtendedView/> component', () => {
    it('renders with proper data', async() => {
        const mocks = [
            {
                request: { query: SINGLE_PRODUCT_QUERY, variables: { id: 'QWERTY0123' } },
                result: {
                    data: {
                        product: fakeProduct(),
                    }
                }
            }
        ];
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ProductExtendedView id='QWERTY0123'/>
            </MockedProvider>
        );

        await act(async() => {
            await new Promise(resolve => setTimeout(resolve, 0));
            wrapper.update();
        });

        // title
        expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
        // image
        expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
        // description
        expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
    });

    it('renders an error when the product is not found', async() => {
        const testID = '321';
        const mocks = [
            {
                request: { query: SINGLE_PRODUCT_QUERY, variables: { id: testID } },
                result: {
                    data: {
                        product: null,
                    }
                }
            }
        ];

        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ProductExtendedView id={testID}/>
            </MockedProvider>
        );

        await act(async() => {
            await new Promise(resolve => setTimeout(resolve, 0));
            wrapper.update();
        });

        expect(wrapper.find('p').text()).toBe(`No Product Found for ${testID}`);
    });
});
