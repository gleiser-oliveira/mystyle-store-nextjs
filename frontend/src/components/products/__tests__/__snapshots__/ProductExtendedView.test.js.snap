// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <ProductExtendedView/> component renders with proper data 1`] = `
<h2>
  Viewing: 
  A Fake Product
</h2>
`;

exports[`Testing <ProductExtendedView/> component renders with proper data 2`] = `
<img
  alt="A Fake Product"
  onLoad={[Function]}
  src="largeProduct.jpg"
/>
`;

exports[`Testing <ProductExtendedView/> component renders with proper data 3`] = `
<p>
  This is a fake product for testing.
</p>
`;
